src/Data/Nat/Primality.agda

```diff
@@ -8,92 +8,186 @@
 open import Relation.Nullary.Decidable as Dec
-  using (yes; no; from-yes; ¬?; decidable-stable; _×-dec_; _→-dec_)
-open import Relation.Nullary.Negation using (¬_; contradiction)
-open import Relation.Unary using (Decidable)
+  using (yes; no; from-yes; from-no; ¬?; _×-dec_; _⊎-dec_; _→-dec_; decidable-stable)
+open import Relation.Nullary.Negation using (¬_; contradiction; contradiction₂)
+open import Relation.Unary using (Pred; Decidable)
+open import Relation.Binary.Core using (Rel)
 open import Relation.Binary.PropositionalEquality
-  using (refl; cong)
+  using (_≡_; _≢_; refl; cong)
 
 private
   variable
-    n : ℕ
+    d m n o p : ℕ
+
+  recompute-nonTrivial : .{{NonTrivial n}} → NonTrivial n
+  recompute-nonTrivial {n} {{nontrivial}} =
+    Dec.recompute (nonTrivial? n) nontrivial
 
 ------------------------------------------------------------------------
 -- Definitions
+------------------------------------------------------------------------
 
--- Definition of compositeness
+-- The definitions of `Prime` and `Composite` in this module are
+-- intended to be built up as complementary pairs of `Decidable`
+-- predicates, with `Prime` given *negatively* as the universal closure
+-- of the negation of `Composite`, where this is given *positively* as
+-- an existential witnessing a non-trivial divisor, where such
+-- quantification is proxied by an explicit `record` type.
 
-Composite : ℕ → Set
-Composite 0 = ⊥
-Composite 1 = ⊥
-Composite n = ∃ λ d → 2 ≤ d × d < n × d ∣ n
+-- The definitions of the predicates `Composite` and `Prime`
+-- as the 'diagonal' instances of relations involving such bounds on
+-- the possible non-trivial divisors leads to the following
+-- positive/existential predicate as the basis for the whole development.
 
--- Definition of primality.
+------------------------------------------------------------------------
+-- Roughness
 
-Prime : ℕ → Set
-Prime 0 = ⊥
-Prime 1 = ⊥
-Prime n = ∀ {d} → 2 ≤ d → d < n → d ∤ n
+-- A number is m-rough if all its non-trivial divisors are bounded below
+-- by m.
+Rough : ℕ → Pred ℕ _
+Rough m n = ¬ (m HasNonTrivialDivisorLessThan n)
 
 ------------------------------------------------------------------------
--- Decidability
+-- Primality
 
-composite? : Decidable Composite
-composite? 0               = no λ()
-composite? 1               = no λ()
-composite? n@(suc (suc _)) = Dec.map′
-  (map₂ λ { (a , b , c) → (b , a , c)})
-  (map₂ λ { (a , b , c) → (b , a , c)})
-  (anyUpTo? (λ d → 2 ≤? d ×-dec d ∣? n) n)
+-- Prime as the diagonal of Rough (and hence the complement of Composite
+-- as defined below). The constructor `prime` takes a proof `isPrime`
+-- that NonTrivial p is p-Rough and thereby enforces that:
+-- * p is a fortiori NonZero and NonUnit
+-- * any non-trivial divisor of p must be at least p, i.e. p itself
+record Prime (p : ℕ) : Set where
+  constructor prime
+  field
+    .{{nontrivial}} : NonTrivial p
+    isPrime         : Rough p p
```

Comment:
I'm a bit worried about the name of this field. It's shouldn't be `isPrime` as otherwise it would be the only field in the record. I'm at a  bit of a loss about what to call it though. Maybe `noDivisors`?

Reply:
Hmm, good thought. I'll ponder that for a bit. Todays a bad day for me, so probably not before tomorrow or the weekend.

Reply:
`isSelfRough`? `isRough`?

Reply:
I could get on board with `isRough`. Or how about `notComposite`?

Reply:
Mond you, if we take this seriously, should we not put the definition of `Compiste` first, then define the type of the field name `noComposite` simply to be `¬ (Composite p)`, with a rider that `Rough p p` is, indeed, definitionally equal to such a property...?

Reply:
Now updated to reflect this.